### ISSUE TITLE
Sockets in CLOSE_WAIT state build up over time

### ISSUE THREAD

**Issue by bcuff** [aspnet/KestrelHttpServer#1785]
After 12 hours or so of running my new asp net core app in production on ubuntu 16.04 I ran out of file handles. lsof showed a ton of sockets in CLOSE_WAIT.

### Minimal repro steps
1. Set up web application on ubuntu 16.04 using kestrel
2. Let it run for a while (exposed to traffic of some kind)
3. Notice that `watch "lsof -n | grep CLOSE_WAIT | wc -l"` builds up slowly over time

### Expected Result
Sockets should be closed properly.

### Actual Result
Sockets hang around forever until the process eventually exceeds its open file limit.

### Further Details

Perhaps this `_socket.Dispose()` call should be moved into the finally block?

See: https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs#L65

```
lsof -n
...
dotnet     1244         www-data  966u     IPv6            5707685       0t0        TCP 127.0.0.1:5000->127.0.0.1:36832 (CLOSE_WAIT)
dotnet     1244         www-data  967u     IPv6            5713986       0t0        TCP 127.0.0.1:5000->127.0.0.1:36450 (CLOSE_WAIT)
dotnet     1244         www-data  968u     IPv6            5712015       0t0        TCP 127.0.0.1:5000->127.0.0.1:36846 (CLOSE_WAIT)
...
```

**Comment by davidfowl** [comment]
Are you running ASP.NET Core 2.0 ci builds in production? If not, your fix does nothing because this code only exists in 2.0. Your socket might be in the CLOSE_WAIT state because you didn't complete your application logic or possibly because of another problem.

**Comment by davidfowl** [comment]
@bcuff Also just FYI Kestrel 1.x isn't suited to be exposed as a webserver without a reverse proxy.

**Comment by bcuff** [comment]
@davidfowl  Okay. I'm using 1.1.1. I guess I was looking in the wrong place.

I am using nginx to front Kestrel. I did turn off proxy_buffering because it was cutting off my responses. That was the only thing different about mine and [this article](https://docs.microsoft.com/en-us/aspnet/core/publishing/linuxproduction)

```
server {
    listen 80;
    listen 8080;
    location / {
        proxy_pass http://localhost:5000;
        proxy_buffering off;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection keep-alive;
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }
```

**Comment by bcuff** [comment]
Ah. I turns out that we had a client that was talking to kestrel directly. When we changed it to point to the nginx proxy instead it stopped leaking sockets. Thanks for the help.

**Comment by mubavoice** [comment]
I am also running the ASP.net core 2.1.1 with SignalR 1.1.0-preview1-34396 in Ubuntu 16.04 I am running out of handlers.

I am running this application behind a Nginx proxy. The asp.net application is running under port 5000


`Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 *:http                  *:*                     LISTEN      795/nginx -g daemon
tcp        0      0 HomeAuto02:domain       *:*                     LISTEN      937/dnsmasq
tcp        0      0 *:ssh                   *:*                     LISTEN      769/sshd
tcp        0      1 192.168.1.51:http       192.168.1.23:6596       LAST_ACK    -
tcp        0      0 192.168.1.51:ssh        SYSLP230.domain.n:51641 ESTABLISHED 23694/sshd: csys
tcp        0    258 localhost:46654         localhost:5000          FIN_WAIT1   -
tcp        0    258 localhost:46666         localhost:5000          FIN_WAIT1   -
tcp        0    255 localhost:46662         localhost:5000          FIN_WAIT1   -
tcp        0      0 192.168.1.51:ssh        SYSLP230.domain.n:52678 ESTABLISHED 24181/sshd: csys
tcp        0    258 localhost:46660         localhost:5000          FIN_WAIT1   -
tcp        0    255 localhost:46668         localhost:5000          FIN_WAIT1   -
tcp        0    258 localhost:46664         localhost:5000          FIN_WAIT1   -
tcp        0    258 localhost:46670         localhost:5000          FIN_WAIT1   -
tcp       32      0 192.168.1.51:33991      117.239.154.134:https   CLOSE_WAIT  2647/dotnet
tcp        0    258 localhost:46656         localhost:5000          FIN_WAIT1   -
tcp        0    255 localhost:46658         localhost:5000          FIN_WAIT1   -
tcp        0    272 192.168.1.51:ssh        SYSLP230.domain.n:51405 ESTABLISHED 23517/sshd: csys
tcp        0      0 192.168.1.51:http       192.168.1.23:9911       SYN_RECV    -
tcp        1      0 192.168.1.51:http       192.168.1.11:20336      CLOSE_WAIT  -
tcp        0    255 localhost:46652         localhost:5000          FIN_WAIT1   -
tcp6       0      0 [::]:5000               [::]:*                  LISTEN      2647/dotnet
tcp6       0      0 [::]:http               [::]:*                  LISTEN      795/nginx -g daemon
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN      769/sshd
tcp6     255      0 localhost:5000          localhost:43156         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43282         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43138         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43314         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43210         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43236         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43288         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43274         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43304         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43212         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43132         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43092         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43258         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43136         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43122         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43320         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43214         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43294         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43280         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43134         CLOSE_WAIT  -
tcp6       0      0 localhost:5000          localhost:46662         SYN_RECV    -
tcp6     258      0 localhost:5000          localhost:43270         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43120         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43208         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43328         CLOSE_WAIT  -
tcp6       1      0 localhost:5000          localhost:43074         CLOSE_WAIT  2647/dotnet
tcp6     255      0 localhost:5000          localhost:43232         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43200         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43182         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43228         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43102         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43128         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43172         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43118         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43202         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43242         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43196         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43154         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43220         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43086         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43204         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43238         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43308         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43250         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43262         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43330         CLOSE_WAIT  -
tcp6       0      0 localhost:5000          localhost:46666         SYN_RECV    -
tcp6     255      0 localhost:5000          localhost:43298         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43170         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43162         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43316         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43226         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43176         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43140         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43322         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43276         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43244         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43142         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43302         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43190         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43318         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43166         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43116         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43084         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43286         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43082         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43192         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43266         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43078         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43110         CLOSE_WAIT  -
tcp6       0      0 localhost:5000          localhost:46670         SYN_RECV    -
tcp6     258      0 localhost:5000          localhost:43254         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43216         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43324         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43184         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43252         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43178         CLOSE_WAIT  -
tcp6       0      0 localhost:5000          localhost:46664         SYN_RECV    -
tcp6     258      0 localhost:5000          localhost:43088         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43234         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43296         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43168         CLOSE_WAIT  -
tcp6       0      0 localhost:5000          localhost:46668         SYN_RECV    -
tcp6     255      0 localhost:5000          localhost:43292         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43144         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43194         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43126         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43096         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43106         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43290         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43336         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43108         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43310         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43180         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43146         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43256         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43264         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43332         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43150         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43306         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43198         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43240         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43174         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43090         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43248         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43224         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43272         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43206         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43312         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43222         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43158         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43188         CLOSE_WAIT  -
tcp6       1      0 localhost:5000          localhost:43076         CLOSE_WAIT  2647/dotnet
tcp6     255      0 localhost:5000          localhost:43112         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43246         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43160         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43260         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43230         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43080         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43094         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43114         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43268         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43152         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43100         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43124         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43278         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43098         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43218         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43300         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43104         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43148         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43326         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43284         CLOSE_WAIT  -
tcp6     258      0 localhost:5000          localhost:43164         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43186         CLOSE_WAIT  -
tcp6     255      0 localhost:5000          localhost:43334         CLOSE_WAIT  -


csysadmin@HomeAuto02:/etc/supervisor/conf.d$ sudo lsof -nP -i
COMMAND     PID         USER   FD   TYPE   DEVICE SIZE/OFF NODE NAME
avahi-dae   646        avahi   12u  IPv4    10584      0t0  UDP *:5353
avahi-dae   646        avahi   13u  IPv6    10585      0t0  UDP *:5353
avahi-dae   646        avahi   14u  IPv4    10586      0t0  UDP *:56497
avahi-dae   646        avahi   15u  IPv6    10587      0t0  UDP *:43010
sshd        769         root    3u  IPv4    14711      0t0  TCP *:22 (LISTEN)
sshd        769         root    4u  IPv6    14713      0t0  TCP *:22 (LISTEN)
nginx       795         root    8u  IPv4    10948      0t0  TCP *:80 (LISTEN)
nginx       795         root    9u  IPv6    10949      0t0  TCP *:80 (LISTEN)
nginx       796     www-data    8u  IPv4    10948      0t0  TCP *:80 (LISTEN)
nginx       796     www-data    9u  IPv6    10949      0t0  TCP *:80 (LISTEN)
nginx       797     www-data    8u  IPv4    10948      0t0  TCP *:80 (LISTEN)
nginx       797     www-data    9u  IPv6    10949      0t0  TCP *:80 (LISTEN)
nginx       798     www-data    8u  IPv4    10948      0t0  TCP *:80 (LISTEN)
nginx       798     www-data    9u  IPv6    10949      0t0  TCP *:80 (LISTEN)
nginx       799     www-data    8u  IPv4    10948      0t0  TCP *:80 (LISTEN)
nginx       799     www-data    9u  IPv6    10949      0t0  TCP *:80 (LISTEN)
dhclient    812         root    6u  IPv4    12576      0t0  UDP *:68
dnsmasq     937       nobody    4u  IPv4    12275      0t0  UDP 127.0.1.1:53
dnsmasq     937       nobody    5u  IPv4    12276      0t0  TCP 127.0.1.1:53 (LISTEN)
dotnet     2647     www-data  296u  IPv6 10196563      0t0  TCP *:5000 (LISTEN)
dotnet     2647     www-data  313u  IPv4 13210569      0t0  TCP 192.168.1.51:33991->117.239.154.134:443 (CLOSE_WAIT)
dotnet     2647     www-data  347u  IPv6 13425828      0t0  TCP 127.0.0.1:5000->127.0.0.1:43074 (CLOSE_WAIT)
dotnet     2647     www-data  349u  IPv6 13425830      0t0  TCP 127.0.0.1:5000->127.0.0.1:43076 (CLOSE_WAIT)
cups-brow 14167         root    8u  IPv4 13430519      0t0  UDP *:631
sshd      23517         root    3u  IPv4 13795081      0t0  TCP 192.168.1.51:22->192.168.1.58:51405 (ESTABLISHED)
sshd      23585    csysadmin    3u  IPv4 13795081      0t0  TCP 192.168.1.51:22->192.168.1.58:51405 (ESTABLISHED)
sshd      23694         root    3u  IPv4 13795165      0t0  TCP 192.168.1.51:22->192.168.1.58:51641 (ESTABLISHED)
sshd      23731    csysadmin    3u  IPv4 13795165      0t0  TCP 192.168.1.51:22->192.168.1.58:51641 (ESTABLISHED)
sshd      24181         root    3u  IPv4 17965881      0t0  TCP 192.168.1.51:22->192.168.1.58:52678 (ESTABLISHED)
sshd      24215    csysadmin    3u  IPv4 17965881      0t0  TCP 192.168.1.51:22->192.168.1.58:52678 (ESTABLISHED)
`

**Comment by mubavoice** [comment]
Any updates on this?

**Comment by davidfowl** [comment]
No there are no updates, we are not actively investigating this. Are these long running websocket connections? How do you reproduce this?

**Comment by lepinay** [comment]
Having a similar issue, I have an asp.net core Background Service running on Amazon Linux that regulary download files from S3 for analysis. 
My first implementation was to use a static WebClient that was reused, and I started to have a lot of CLOSE_WAIT pilling up after a few hours.
I changed the implementation to use a WebClient in a using() block for each download, this improved the situation but the issue just moved from hours to days.

I'm going now to try using HttpClient instead to see if that changes anything....
